vortex-mask/src/lib.rs

```diff
@@ -7,104 +7,151 @@ mod intersect_by_rank;
 mod iter_bools;
 
 use std::cmp::Ordering;
+use std::fmt::{Debug, Formatter};
+use std::mem::discriminant;
 use std::sync::{Arc, OnceLock};
 
 use arrow_buffer::{BooleanBuffer, BooleanBufferBuilder};
 use itertools::Itertools;
 use vortex_error::vortex_panic;
 
-/// If the mask selects more than this fraction of rows, iterate over slices instead of indices.
-///
-/// Threshold of 0.8 chosen based on Arrow Rust, which is in turn based on:
-///   <https://dl.acm.org/doi/abs/10.1145/3465998.3466009>
-const FILTER_SLICES_SELECTIVITY_THRESHOLD: f64 = 0.8;
+/// Represents a set of values that are all included, all excluded, or some mixture of both.
+pub enum AllOr<T> {
+    /// All values are included.
+    All,
+    /// No values are included.
+    None,
+    /// Some values are included.
+    Some(T),
+}
+
+impl<T> AllOr<T> {
+    /// Returns the `Some` variant of the enum.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the variant is not `Some`.
+    pub fn expect_some(self) -> T {
+        match self {
+            Self::Some(v) => v,
+            _ => vortex_panic!("Expected Some variant, got {:?}", discriminant(&self)),
+        }
+    }
+
+    /// Returns the `Some` variant of the enum, or a default value.
+    pub fn unwrap_or_else<F, G>(self, all_true: F, all_false: G) -> T
+    where
+        F: FnOnce() -> T,
+        G: FnOnce() -> T,
+    {
+        match self {
+            Self::Some(v) => v,
+            AllOr::All => all_true(),
+            AllOr::None => all_false(),
+        }
+    }
+}
+
+impl<T> AllOr<&T> {
+    /// Clone the inner value.
+    pub fn cloned(self) -> AllOr<T>
+    where
+        T: Clone,
+    {
+        match self {
+            Self::All => AllOr::All,
+            Self::None => AllOr::None,
+            Self::Some(v) => AllOr::Some(v.clone()),
+        }
+    }
+}
+
+impl<T> Debug for AllOr<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::All => f.write_str("All"),
+            Self::None => f.write_str("None"),
+            Self::Some(v) => f.debug_tuple("Some").field(v).finish(),
+        }
+    }
+}
+
+impl<T> PartialEq for AllOr<T>
+where
+    T: PartialEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::All, Self::All) => true,
+            (Self::None, Self::None) => true,
+            (Self::Some(lhs), Self::Some(rhs)) => lhs == rhs,
+            _ => false,
+        }
+    }
+}
+
+impl<T> Eq for AllOr<T> where T: Eq {}
 
 /// Represents a set of sorted unique positive integers.
 ///
 /// A [`Mask`] can be constructed from various representations, and converted to various
 /// others. Internally, these are cached.
 #[derive(Clone, Debug)]
-pub struct Mask(pub(crate) Arc<Inner>);
+pub enum Mask {
```

Comment:
I don't think the enum should be exposed, we already had the `new_true` and `new_false` constructors previously, why not just continue to use those and have them delegate to the specialized enum storage?